File: Leetcode/generateParentheses.js


// https://leetcode.com/problems/generate-parentheses/description/

// Solution

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n, memo) {
    if (n === 0) return [''];
    if (!memo) memo = new Map();
    if (memo.has(n)) return memo.get(n);

    const retArr = [];
    for (let i = 0; i < n; i++) {
        const leftArr = generateParenthesis(i, memo);
        const rightArr = generateParenthesis(n - i - 1, memo);
        leftArr.forEach(leftParen => {
            rightArr.forEach(rightParen => {
                retArr.push('(' + leftParen + ')' + rightParen);
            });
        });
    }

    memo.set(n, retArr);
    return retArr;
};

// Run: O(4^n / sqrt{n}})  - not sure why.. (Catalan number's upper bound??)
// Space: O(4^n / sqrt{n}})

File: Leetcode/recursion & dp/generateParentheses.cs


// https://leetcode.com/problems/generate-parentheses/description/

public class Solution {
    public List<string> GenerateParenthesis(int n) {
        Dictionary<int, List<string>> memo = new Dictionary<int, List<string>>();
        return GenerateParenthesis(n, memo);
    }

    public List<string> GenerateParenthesis(int n, Dictionary<int, List<string>> memo) {
        List<string> retList = new List<string>();
        if (n == 0) {
          retList.Add("");
          return retList;
        }
        if (memo.ContainsKey(n)) return memo[n];

        for (int i = 0; i < n; i++) {
          List<string> leftList = GenerateParenthesis(i, memo);
          List<string> rightList = GenerateParenthesis(n - i - 1, memo);
          foreach(string left in leftList) {
            foreach(string right in rightList) {
              retList.Add("(" + left + ")" + right);
            }
          }
        }
        memo.Add(n, retList);
        return retList;
    }
}

// Run: O(4^n / sqrt{n}})  - not sure why.. (Catalan number's upper bound??)
// Space: O(4^n / sqrt{n}})
